libraries/AP_TECS/AP_TECS.cpp

```diff
@@ -1436,3 +1420,82 @@ void AP_TECS::update_pitch_throttle(int32_t hgt_dem_cm,
     }
 #endif
 }
+
+// set minimum throttle override, [-1, -1] range
+// it is applicable for one control cycle only
+void AP_TECS::set_throttle_min(const float thr_min) {
+    // Don't change the limit if it is already covered.
+    if (thr_min > _THRminf_ext) {
+        _THRminf_ext = thr_min;
+    }
+}
+
+// set minimum throttle override, [0, -1] range
+// it is applicable for one control cycle only
+void AP_TECS::set_throttle_max(const float thr_max) {
+    // Don't change the limit if it is already covered.
+    if (thr_max < _THRmaxf_ext) {
+        _THRmaxf_ext = thr_max;
+    }
+}
+
+void AP_TECS::_update_throttle_limits() {
+
+    // Configure max throttle.
+
+    // Read the maximum throttle limit.
+    _THRmaxf  = aparm.throttle_max * 0.01f;
+    // If more max throttle is allowed during takeoff, use it.
+    if (aparm.takeoff_throttle_max*0.01f > _THRmaxf
+        && (_flight_stage == AP_FixedWing::FlightStage::TAKEOFF || _flight_stage == AP_FixedWing::FlightStage::ABORT_LANDING)
+        ) {
+        _THRmaxf  = aparm.takeoff_throttle_max * 0.01f;
+    }
+    // In any case, constrain to the external safety limits.
+    _THRmaxf = MIN(_THRmaxf, _THRmaxf_ext);
+
+    // Configure min throttle.
+
+    // If less min throttle is allowed during takeoff, use it.
+    bool use_takeoff_throttle = _flight_stage == AP_FixedWing::FlightStage::TAKEOFF || _flight_stage == AP_FixedWing::FlightStage::ABORT_LANDING;
+    use_takeoff_throttle = use_takeoff_throttle && (aparm.takeoff_mode == 1) && (aparm.takeoff_throttle_min != 0);
+    if ( use_takeoff_throttle ) { 
+        _THRminf = MIN(_THRminf, aparm.takeoff_throttle_min * 0.01f);
+    }
+    else { // Otherwise, during normal situations let regular limit.
+        _THRminf = aparm.throttle_min * 0.01f;
+    }
+    // Raise min to force max throttle for TKOFF_THR_MAX_T after a takeoff.
+    if (_flight_stage == AP_FixedWing::FlightStage::TAKEOFF) {
+        const uint32_t now = AP_HAL::millis();
+        if (_takeoff_start_ms == 0) {
+            _takeoff_start_ms = now;
+        }
+        const uint32_t dt = now - _takeoff_start_ms;
+        if (dt*0.001 < aparm.takeoff_throttle_max_t) {
+            _THRminf = _THRmaxf;
+        } else {
```

Comment:
what is this????

Reply:
This particular if-statement is the implementation of TKOFF_THR_MAX_T: It ensures throttle is at max for the first TKOFF_THR_MAX_T seconds.

I didn't create it; I moved it from `_update_throttle_without_airspeed` further above to make it common with both with- and without-airspeed measurement cases. Thus simplifying the default takeoff logic.

Reply:
Geroge, the i variable is unused